In [1]:
import sys
sys.path.append('..')
from src.kitti import *
import numpy as np
import cv2
import json
import glob
from src.boundingbox import format_yolo_output
from src.trajectory import Trajectory
from src.sequence import load_keyframes

%matplotlib inline

### Analyzing KITTI sequence 00 image_2

In [ ]:
LEFTDIR = '/data/urbinn/datasets/kitti/sequences/00/image_0'
RIGHTDIR = '/data/urbinn/datasets/kitti/sequences/00/image_1'

yolo_data = format_yolo_output("/data/urbinn/darknet/output/seq00_image02/objects.json")
                
frames = []
for filename in sorted(glob.glob(LEFTDIR + '/*')): 
    image_name = filename.split('/')[-1]
    frame = None
    
    if not image_name in yolo_data:
        print("frame {} has no yolo data available".format(image_name))
        frame = Frame(filename, RIGHTDIR)
    else:
        frame = Frame(filename, RIGHTDIR, classifications = yolo_data[image_name])
        
    frames.append(frame)

frame 004540.png has no yolo data available


Note: For some reason frame 004540 has no objects detected. This might be possible but requires further investigation.

#### Setting up a sequence

In [ ]:
seq = Sequence()
for f in ProgressBar()(frames):
    seq.add_keyframe(f)

  0% (1 of 4541) |                        | Elapsed Time: 0:00:00 ETA:  0:50:31

connected mappoints  0
1 
 [[ 0.99999083  0.00273335  0.00329735  0.00518481]
 [-0.00274001  0.99999421  0.00201826  0.00702947]
 [-0.00329182 -0.00202728  0.99999253 -0.67054258]
 [ 0.          0.          0.          1.        ]]


  0% (2 of 4541) |                        | Elapsed Time: 0:00:01 ETA:  1:04:10

connected mappoints  83
2 
 [[ 0.99999134 -0.00135221  0.00393651  0.01068681]
 [ 0.00134508  0.99999745  0.00181499 -0.00406298]
 [-0.00393896 -0.00180968  0.9999906  -0.68970785]
 [ 0.          0.          0.          1.        ]]


  0% (3 of 4541) |                        | Elapsed Time: 0:00:02 ETA:  1:08:29

connected mappoints  98
3 
 [[  9.99991764e-01  -1.84997578e-04   4.05445531e-03   5.83343991e-03]
 [  1.78884939e-04   9.99998847e-01   1.50794580e-03   9.91494890e-04]
 [ -4.05472961e-03  -1.50720810e-03   9.99990644e-01  -7.15266115e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  0% (4 of 4541) |                        | Elapsed Time: 0:00:03 ETA:  1:11:51

connected mappoints  102
4 
 [[ 0.9999853  -0.00135978  0.00524797  0.00425461]
 [ 0.00135442  0.99999856  0.00102434 -0.0047628 ]
 [-0.00524936 -0.00101722  0.9999857  -0.72796975]
 [ 0.          0.          0.          1.        ]]


  0% (5 of 4541) |                        | Elapsed Time: 0:00:04 ETA:  1:13:19

connected mappoints  90
5 
 [[  9.99990198e-01   1.63048170e-03   4.11657080e-03   1.58865776e-02]
 [ -1.62903460e-03   9.99998610e-01  -3.54858755e-04   1.42036162e-02]
 [ -4.11714367e-03   3.48149240e-04   9.99991464e-01  -7.23474087e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  0% (6 of 4541) |                        | Elapsed Time: 0:00:05 ETA:  1:14:22

connected mappoints  94
6 
 [[  9.99988806e-01  -1.69069774e-03   4.41920712e-03   9.59916731e-03]
 [  1.68487823e-03   9.99997709e-01   1.32025998e-03   6.82636474e-04]
 [ -4.42142915e-03  -1.31279937e-03   9.99989364e-01  -7.52441246e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  0% (7 of 4541) |                        | Elapsed Time: 0:00:06 ETA:  1:14:22

connected mappoints  98
7 
 [[  9.99985830e-01  -2.15298070e-03   4.86873512e-03   1.24665209e-02]
 [  2.15012464e-03   9.99997513e-01   5.91770581e-04   2.11096979e-03]
 [ -4.86999709e-03  -5.81293808e-04   9.99987973e-01  -7.70046244e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  0% (8 of 4541) |                        | Elapsed Time: 0:00:07 ETA:  1:15:02

connected mappoints  102
8 
 [[  9.99989937e-01  -5.37581893e-04   4.45385360e-03   5.03043897e-03]
 [  5.44055168e-04   9.99998797e-01  -1.45232602e-03   6.24195607e-04]
 [ -4.45306749e-03   1.45473455e-03   9.99989027e-01  -7.91745414e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  0% (9 of 4541) |                        | Elapsed Time: 0:00:08 ETA:  1:14:50

connected mappoints  107
9 
 [[  9.99990180e-01   7.79901665e-04   4.36263695e-03   8.59239841e-03]
 [ -7.68178264e-04   9.99996092e-01  -2.68825878e-03   8.09069197e-03]
 [ -4.36471647e-03   2.68488110e-03   9.99986870e-01  -8.00163029e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  0% (10 of 4541) |                       | Elapsed Time: 0:00:09 ETA:  1:14:00

connected mappoints  87
10 
 [[  9.99995150e-01  -5.16391866e-05   3.11412754e-03   1.18601698e-02]
 [  5.92000738e-05   9.99997051e-01  -2.42788784e-03   8.69088612e-03]
 [ -3.11399298e-03   2.42806042e-03   9.99992204e-01  -7.98032084e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  0% (11 of 4541) |                       | Elapsed Time: 0:00:10 ETA:  1:12:27

connected mappoints  87
11 
 [[  9.99995133e-01  -1.68043107e-03   2.62870369e-03   7.40091709e-03]
 [  1.67835766e-03   9.99998279e-01   7.90765774e-04   8.55980375e-03]
 [ -2.63002799e-03  -7.86350020e-04   9.99996232e-01  -8.17197797e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  0% (12 of 4541) |                       | Elapsed Time: 0:00:11 ETA:  1:10:46

connected mappoints  91
12 
 [[  9.99997184e-01   5.04614666e-04   2.31905974e-03   1.26955486e-03]
 [ -5.05129903e-04   9.99999848e-01   2.21594530e-04   5.80065505e-03]
 [ -2.31894757e-03  -2.22765332e-04   9.99997286e-01  -8.25465935e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  0% (13 of 4541) |                       | Elapsed Time: 0:00:11 ETA:  1:09:15

connected mappoints  114
13 
 [[  9.99994355e-01   2.77119130e-03   1.90034779e-03  -1.57598289e-03]
 [ -2.77286159e-03   9.99995771e-01   8.76865748e-04  -8.43203999e-04]
 [ -1.89790979e-03  -8.82130199e-04   9.99997810e-01  -8.39318606e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  0% (14 of 4541) |                       | Elapsed Time: 0:00:12 ETA:  1:07:57

connected mappoints  92
14 
 [[  9.99997299e-01   1.05617817e-03   2.07040089e-03   2.14753562e-04]
 [ -1.05534202e-03   9.99999361e-01  -4.04908256e-04  -1.06837323e-02]
 [ -2.07082723e-03   4.02722181e-04   9.99997775e-01  -8.30516592e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  0% (15 of 4541) |                       | Elapsed Time: 0:00:13 ETA:  1:06:41

connected mappoints  104
15 
 [[  9.99997336e-01   1.78097870e-03   1.46858799e-03   6.15186445e-03]
 [ -1.78094562e-03   9.99998414e-01  -2.38300807e-05   1.51770006e-02]
 [ -1.46862811e-03   2.12145419e-05   9.99998921e-01  -8.61730077e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  0% (16 of 4541) |                       | Elapsed Time: 0:00:13 ETA:  1:05:37

connected mappoints  103
16 
 [[ 0.99999573 -0.00230442  0.00179848  0.00754146]
 [ 0.00230014  0.99999453  0.00237724  0.00300301]
 [-0.00180395 -0.00237309  0.99999556 -0.85114646]
 [ 0.          0.          0.          1.        ]]


  0% (17 of 4541) |                       | Elapsed Time: 0:00:14 ETA:  1:04:36

connected mappoints  100
17 
 [[  9.99999106e-01  -4.26566834e-04   1.26734789e-03   7.06255911e-03]
 [  4.23146496e-04   9.99996271e-01   2.69785877e-03   3.93827169e-03]
 [ -1.26849398e-03  -2.69732008e-03   9.99995558e-01  -8.70569552e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  0% (18 of 4541) |                       | Elapsed Time: 0:00:15 ETA:  1:03:51

connected mappoints  85
18 
 [[ 0.99999273 -0.00330122  0.00190895  0.00804218]
 [ 0.00329845  0.99999351  0.00145082  0.0030749 ]
 [-0.00191372 -0.00144452  0.99999713 -0.87341901]
 [ 0.          0.          0.          1.        ]]


  0% (19 of 4541) |                       | Elapsed Time: 0:00:16 ETA:  1:03:29

connected mappoints  92
19 
 [[  9.99997535e-01  -2.06081062e-03   8.26236531e-04   1.31562966e-02]
 [  2.06092341e-03   9.99997867e-01  -1.35677175e-04   9.14846820e-03]
 [ -8.25955164e-04   1.37379650e-04   9.99999649e-01  -8.89061157e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  0% (20 of 4541) |                       | Elapsed Time: 0:00:16 ETA:  1:03:19

connected mappoints  97
20 
 [[  9.99999235e-01   5.90422084e-04   1.08671717e-03   1.04595715e-02]
 [ -5.89319595e-04   9.99999312e-01  -1.01455378e-03   3.99454134e-03]
 [ -1.08731544e-03   1.01391258e-03   9.99998895e-01  -8.83705692e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  0% (21 of 4541) |                       | Elapsed Time: 0:00:17 ETA:  1:03:14

connected mappoints  101
21 
 [[  9.99998950e-01  -4.89902237e-04   1.36390185e-03   4.46499570e-03]
 [  4.91555459e-04   9.99999145e-01  -1.21205563e-03   3.46735154e-03]
 [ -1.36330689e-03   1.21272479e-03   9.99998335e-01  -8.98477580e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  0% (22 of 4541) |                       | Elapsed Time: 0:00:18 ETA:  1:03:34

connected mappoints  131
22 
 [[  9.99999455e-01   7.62652063e-04   7.12446695e-04   1.04169538e-02]
 [ -7.62578886e-04   9.99999704e-01  -1.02977851e-04   6.54235390e-03]
 [ -7.12525021e-04   1.02434498e-04   9.99999741e-01  -8.96437845e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  0% (23 of 4541) |                       | Elapsed Time: 0:00:19 ETA:  1:04:11

connected mappoints  130
23 
 [[  9.99999094e-01   1.48801855e-05   1.34590553e-03   1.04658068e-02]
 [ -1.60627480e-05   9.99999614e-01   8.78630491e-04   7.46124846e-03]
 [ -1.34589194e-03  -8.78651314e-04   9.99998708e-01  -9.01721538e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  0% (24 of 4541) |                       | Elapsed Time: 0:00:20 ETA:  1:04:55

connected mappoints  115
24 
 [[  9.99998720e-01   1.43630641e-03   7.05268914e-04   2.81350344e-02]
 [ -1.43678361e-03   9.99998739e-01   6.76579023e-04   8.70283473e-03]
 [ -7.04296250e-04  -6.77591476e-04   9.99999522e-01  -9.46884355e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  0% (25 of 4541) |                       | Elapsed Time: 0:00:21 ETA:  1:05:30

connected mappoints  104
25 
 [[  9.99999767e-01   3.75681013e-04   5.70119794e-04   3.95796414e-04]
 [ -3.74645526e-04   9.99998282e-01  -1.81528309e-03   7.22370995e-03]
 [ -5.70800782e-04   1.81506907e-03   9.99998190e-01  -9.17229749e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  0% (26 of 4541) |                       | Elapsed Time: 0:00:22 ETA:  1:05:58

connected mappoints  101
26 
 [[  9.99998699e-01  -1.57592410e-03  -3.44462215e-04   1.42178994e-02]
 [  1.57467871e-03   9.99992329e-01  -3.58630684e-03   1.92585976e-02]
 [  3.50111320e-04   3.58575976e-03   9.99993510e-01  -9.22576438e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  0% (27 of 4541) |                       | Elapsed Time: 0:00:23 ETA:  1:06:24

connected mappoints  115
27 
 [[  9.99999626e-01  -4.13664832e-04  -7.59174408e-04   1.59600709e-03]
 [  4.11013285e-04   9.99993827e-01  -3.48951072e-03   5.09186597e-03]
 [  7.60613210e-04   3.48919738e-03   9.99993623e-01  -9.16668738e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  0% (28 of 4541) |                       | Elapsed Time: 0:00:24 ETA:  1:06:43

connected mappoints  103
28 
 [[ 0.99999766  0.0013307  -0.00170794  0.00475206]
 [-0.00133569  0.99999484 -0.00292046  0.01539057]
 [ 0.00170405  0.00292273  0.99999428 -0.91780441]
 [ 0.          0.          0.          1.        ]]


  0% (29 of 4541) |                       | Elapsed Time: 0:00:25 ETA:  1:07:06

connected mappoints  115
29 
 [[  9.99998809e-01   4.87083701e-05  -1.54276358e-03   3.50763040e-03]
 [ -4.98330489e-05   9.99999733e-01  -7.28972692e-04   7.55416457e-03]
 [  1.54272766e-03   7.29048704e-04   9.99998544e-01  -9.41365100e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  0% (30 of 4541) |                       | Elapsed Time: 0:00:26 ETA:  1:07:13

connected mappoints  89
30 
 [[ 0.99999792 -0.0016244  -0.00123679 -0.00296523]
 [ 0.00162779  0.99999491  0.00274422  0.01919504]
 [ 0.00123233 -0.00274623  0.99999547 -0.92733107]
 [ 0.          0.          0.          1.        ]]


  0% (31 of 4541) |                       | Elapsed Time: 0:00:27 ETA:  1:06:53

connected mappoints  97
31 
 [[ 0.99999881 -0.00116187 -0.00101199  0.00965927]
 [ 0.00116688  0.999987    0.00496386  0.00291027]
 [ 0.00100621 -0.00496503  0.99998717 -0.95078526]
 [ 0.          0.          0.          1.        ]]


  0% (32 of 4541) |                       | Elapsed Time: 0:00:28 ETA:  1:06:55

connected mappoints  105
32 
 [[  9.99998924e-01   3.59234400e-04  -1.42222514e-03   3.97064198e-03]
 [ -3.55132940e-04   9.99995781e-01   2.88303649e-03   8.62108066e-03]
 [  1.42325482e-03  -2.88252831e-03   9.99994833e-01  -9.49914328e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  0% (33 of 4541) |                       | Elapsed Time: 0:00:29 ETA:  1:07:00

connected mappoints  103
33 
 [[  9.99996777e-01   2.44241028e-03  -6.92596840e-04   4.61783380e-03]
 [ -2.44298542e-03   9.99996671e-01  -8.30776735e-04   3.19670436e-03]
 [  6.90565437e-04   8.32466062e-04   9.99999415e-01  -9.75129885e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  0% (34 of 4541) |                       | Elapsed Time: 0:00:30 ETA:  1:07:05

connected mappoints  96
34 
 [[  9.99999212e-01   1.25534410e-03   2.82801746e-05   4.36769267e-03]
 [ -1.25518944e-03   9.99987063e-01  -4.92941453e-03   1.46747175e-02]
 [ -3.44679202e-05   4.92937514e-03   9.99987850e-01  -9.58659965e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  0% (35 of 4541) |                       | Elapsed Time: 0:00:31 ETA:  1:07:21

connected mappoints  140
35 
 [[  9.99991727e-01   3.99241170e-03   7.78717620e-04  -4.31518202e-03]
 [ -3.98984062e-03   9.99986675e-01  -3.27574692e-03   1.21261080e-02]
 [ -7.91785374e-04   3.27261286e-03   9.99994332e-01  -9.62986298e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  0% (36 of 4541) |                       | Elapsed Time: 0:00:32 ETA:  1:07:43

connected mappoints  130
36 
 [[  9.99969080e-01   7.81869615e-03   8.40278139e-04  -1.80370422e-03]
 [ -7.81917709e-03   9.99969267e-01   5.70597638e-04   8.15419323e-03]
 [ -8.35790985e-04  -5.77150279e-04   9.99999484e-01  -9.56091458e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  0% (37 of 4541) |                       | Elapsed Time: 0:00:33 ETA:  1:07:52

connected mappoints  91
37 
 [[  9.99996734e-01   2.20302242e-03   1.29588527e-03   6.65080484e-03]
 [ -2.20820783e-03   9.99989507e-01   4.01371850e-03   3.09908210e-04]
 [ -1.28702936e-03  -4.01656697e-03   9.99991105e-01  -9.62949103e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  0% (38 of 4541) |                       | Elapsed Time: 0:00:34 ETA:  1:07:55

connected mappoints  94
38 
 [[  9.99998966e-01  -5.33369817e-04   1.33531022e-03   7.11811752e-03]
 [  5.28124447e-04   9.99992156e-01   3.92547877e-03   3.85087219e-03]
 [ -1.33739348e-03  -3.92476950e-03   9.99991404e-01  -9.71898034e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  0% (39 of 4541) |                       | Elapsed Time: 0:00:35 ETA:  1:08:12

connected mappoints  109
39 
 [[  9.99996397e-01  -1.65213921e-03   2.11593166e-03  -1.01273544e-03]
 [  1.65064435e-03   9.99998387e-01   7.08030275e-04   1.18620277e-02]
 [ -2.11709801e-03  -7.04535073e-04   9.99997511e-01  -9.59307775e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  0% (40 of 4541) |                       | Elapsed Time: 0:00:36 ETA:  1:08:24

connected mappoints  109
40 
 [[  9.99999678e-01   4.71713506e-04   6.49574742e-04   9.12772403e-03]
 [ -4.70803222e-04   9.99998908e-01  -1.40079507e-03   6.26355524e-03]
 [ -6.50234806e-04   1.40048880e-03   9.99998808e-01  -9.87937039e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  0% (41 of 4541) |                       | Elapsed Time: 0:00:37 ETA:  1:08:17

connected mappoints  90
41 
 [[ 0.99999536 -0.00279385  0.00121571  0.00742597]
 [ 0.0027972   0.99999229 -0.00275564  0.01128243]
 [-0.001208    0.00275902  0.99999546 -0.98329348]
 [ 0.          0.          0.          1.        ]]


  0% (42 of 4541) |                       | Elapsed Time: 0:00:38 ETA:  1:08:08

connected mappoints  91
42 
 [[  9.99999006e-01   4.25956472e-04   1.34381662e-03   3.65686978e-03]
 [ -4.24254580e-04   9.99999108e-01  -1.26649222e-03   1.23070388e-02]
 [ -1.34435489e-03   1.26592084e-03   9.99998295e-01  -1.00406060e+00]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  0% (43 of 4541) |                       | Elapsed Time: 0:00:39 ETA:  1:08:03

connected mappoints  95
43 
 [[  9.99998674e-01  -1.37122831e-03   8.78199576e-04   3.24231839e-03]
 [  1.37350012e-03   9.99995699e-01  -2.59153947e-03   1.44504723e-02]
 [ -8.74642206e-04   2.59274224e-03   9.99996256e-01  -9.70842511e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  0% (44 of 4541) |                       | Elapsed Time: 0:00:39 ETA:  1:07:58

connected mappoints  107
44 
 [[  9.99999929e-01   3.28808430e-04  -1.83304973e-04   3.05513837e-03]
 [ -3.29308054e-04   9.99996213e-01  -2.73230673e-03   8.01195737e-03]
 [  1.82405873e-04   2.73236690e-03   9.99996250e-01  -9.90065774e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  0% (45 of 4541) |                       | Elapsed Time: 0:00:40 ETA:  1:08:22

connected mappoints  120
45 
 [[  9.99984696e-01  -5.49947828e-03   6.02835083e-04   1.14553372e-02]
 [  5.50090301e-03   9.99982019e-01  -2.38775892e-03   2.17678408e-02]
 [ -5.89692815e-04   2.39103851e-03   9.99996968e-01  -9.79483869e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  1% (46 of 4541) |                       | Elapsed Time: 0:00:41 ETA:  1:08:04

connected mappoints  102
46 
 [[  9.99999778e-01   6.41665119e-04   1.80627996e-04   1.33491898e-03]
 [ -6.41750176e-04   9.99999683e-01   4.71229919e-04   1.26125876e-02]
 [ -1.80325567e-04  -4.71345732e-04   9.99999873e-01  -9.85162811e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  1% (47 of 4541) |                       | Elapsed Time: 0:00:42 ETA:  1:07:40

connected mappoints  96
47 
 [[  9.99999884e-01  -4.80783080e-04  -1.20158899e-05   9.02092967e-04]
 [  4.80802627e-04   9.99998467e-01   1.68347690e-03   1.37725189e-02]
 [  1.12064843e-05  -1.68348248e-03   9.99998583e-01  -9.97598826e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  1% (48 of 4541) |                       | Elapsed Time: 0:00:43 ETA:  1:07:16

connected mappoints  87
48 
 [[  9.99997735e-01  -2.12837699e-03  -2.95133578e-05   8.70656720e-03]
 [  2.12844381e-03   9.99994642e-01   2.48714642e-03   2.89982447e-03]
 [  2.42196144e-05  -2.48720361e-03   9.99996907e-01  -9.86934221e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  1% (49 of 4541) |                       | Elapsed Time: 0:00:44 ETA:  1:06:51

connected mappoints  105
49 
 [[  9.99999254e-01   7.45239810e-04   9.67376472e-04   3.61855356e-03]
 [ -7.45101539e-04   9.99999712e-01  -1.43286681e-04   1.46695217e-02]
 [ -9.67482977e-04   1.42565781e-04   9.99999522e-01  -9.86888317e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  1% (50 of 4541) |                       | Elapsed Time: 0:00:44 ETA:  1:06:19

connected mappoints  97
50 
 [[ 0.99999548  0.00256839  0.00156115 -0.01140818]
 [-0.00256675  0.99999615 -0.00105115  0.00669282]
 [-0.00156384  0.00104714  0.99999823 -0.9822767 ]
 [ 0.          0.          0.          1.        ]]


  1% (51 of 4541) |                       | Elapsed Time: 0:00:45 ETA:  1:05:52

connected mappoints  90
51 
 [[  9.99999682e-01   2.13686253e-04   7.68904284e-04   1.60753526e-02]
 [ -2.12979259e-04   9.99999555e-01  -9.19446303e-04   3.14679381e-03]
 [ -7.69100414e-04   9.19282249e-04   9.99999282e-01  -9.87004901e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  1% (52 of 4541) |                       | Elapsed Time: 0:00:46 ETA:  1:05:15

connected mappoints  83
52 
 [[  9.99997018e-01  -2.43447617e-03   1.91308334e-04   1.14555079e-02]
 [  2.43378126e-03   9.99990729e-01   3.55238202e-03   1.31740011e-02]
 [ -1.99954749e-04  -3.55190582e-03   9.99993672e-01  -9.59730056e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  1% (53 of 4541) |                       | Elapsed Time: 0:00:47 ETA:  1:04:46

connected mappoints  101
53 
 [[  9.99999569e-01  -6.38676118e-04   6.74009359e-04   1.45699931e-02]
 [  6.35828779e-04   9.99990909e-01   4.21627282e-03   7.30474005e-03]
 [ -6.76696064e-04  -4.21584245e-03   9.99990884e-01  -9.59685866e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  1% (54 of 4541) |                       | Elapsed Time: 0:00:47 ETA:  1:04:29

connected mappoints  100
54 
 [[  9.99999662e-01  -7.72458218e-04   2.80381725e-04   1.02510385e-02]
 [  7.72075178e-04   9.99998772e-01   1.36368428e-03   8.01490114e-03]
 [ -2.81434770e-04  -1.36346734e-03   9.99999031e-01  -9.61003394e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  1% (55 of 4541) |                       | Elapsed Time: 0:00:48 ETA:  1:04:27

connected mappoints  87
55 
 [[  9.99997934e-01   2.02768864e-03   1.40740762e-04   8.24900779e-03]
 [ -2.02720421e-03   9.99992298e-01  -3.36076416e-03   1.32663465e-02]
 [ -1.47554261e-04   3.36047191e-03   9.99994343e-01  -9.81238866e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  1% (56 of 4541) |                       | Elapsed Time: 0:00:49 ETA:  1:04:29

connected mappoints  81
56 
 [[  9.99997484e-01   2.24245243e-03  -5.11820102e-05   8.02739450e-03]
 [ -2.24258934e-03   9.99993415e-01  -2.85313132e-03   9.62904368e-03]
 [  4.47836619e-05   2.85323892e-03   9.99995929e-01  -9.72464967e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  1% (57 of 4541) |                       | Elapsed Time: 0:00:49 ETA:  1:04:35

connected mappoints  73
57 
 [[  9.99997964e-01   2.00575546e-03   2.21870553e-04   4.31029437e-03]
 [ -2.00560722e-03   9.99997767e-01  -6.66353660e-04   1.36362745e-02]
 [ -2.23206600e-04   6.65907318e-04   9.99999753e-01  -9.54722147e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  1% (58 of 4541) |                       | Elapsed Time: 0:00:50 ETA:  1:04:51

connected mappoints  103
58 
 [[  9.99994418e-01   3.34080951e-03   6.01231645e-05   3.44554053e-03]
 [ -3.34087372e-03   9.99993813e-01   1.10153455e-03   1.42039036e-02]
 [ -5.64427754e-05  -1.10172927e-03   9.99999392e-01  -9.53721083e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  1% (59 of 4541) |                       | Elapsed Time: 0:00:51 ETA:  1:05:19

connected mappoints  110
59 
 [[  9.99996733e-01   2.45599892e-03   7.08681043e-04  -3.98419089e-03]
 [ -2.45691517e-03   9.99996143e-01   1.29493696e-03   1.60901496e-02]
 [ -7.05497946e-04  -1.29667390e-03   9.99998910e-01  -9.66172181e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  1% (60 of 4541) |                       | Elapsed Time: 0:00:52 ETA:  1:05:39

connected mappoints  96
60 
 [[  9.99998065e-01   1.77659509e-03   8.44485286e-04   7.89753214e-03]
 [ -1.77560989e-03   9.99997744e-01  -1.16594810e-03   1.01627200e-02]
 [ -8.46554798e-04   1.16444636e-03   9.99998964e-01  -9.66442985e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  1% (61 of 4541) |                       | Elapsed Time: 0:00:53 ETA:  1:05:57

connected mappoints  104
61 
 [[ 0.99999447  0.0031271   0.00112756 -0.00224781]
 [-0.00312245  0.99998671 -0.00410192  0.01797783]
 [-0.00114037  0.00409837  0.99999095 -0.93318251]
 [ 0.          0.          0.          1.        ]]


  1% (62 of 4541) |                       | Elapsed Time: 0:00:54 ETA:  1:06:12

connected mappoints  105
62 
 [[  9.99997867e-01   1.96537329e-03   6.35055493e-04  -7.11011085e-03]
 [ -1.96218932e-03   9.99985694e-01  -4.97600796e-03   1.39082847e-02]
 [ -6.44826122e-04   4.97475124e-03   9.99987418e-01  -9.35937312e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  1% (63 of 4541) |                       | Elapsed Time: 0:00:55 ETA:  1:06:22

connected mappoints  103
63 
 [[  9.99997550e-01  -2.19603602e-03   2.78702537e-04   1.07033821e-02]
 [  2.19669460e-03   9.99994743e-01  -2.38511796e-03   2.55919652e-02]
 [ -2.73463267e-04   2.38572434e-03   9.99997117e-01  -9.36702022e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  1% (64 of 4541) |                       | Elapsed Time: 0:00:55 ETA:  1:06:27

connected mappoints  96
64 
 [[ 0.99999805 -0.00139706  0.00139387  0.00719462]
 [ 0.0013914   0.99999082  0.00405342  0.01103953]
 [-0.00139952 -0.00405147  0.99999081 -0.93128845]
 [ 0.          0.          0.          1.        ]]


  1% (65 of 4541) |                       | Elapsed Time: 0:00:56 ETA:  1:06:29

connected mappoints  114
65 
 [[ 0.99999669 -0.0021817   0.00136306  0.00847938]
 [ 0.00217103  0.99996736  0.00778274  0.0040484 ]
 [-0.00137999 -0.00777975  0.99996879 -0.92261212]
 [ 0.          0.          0.          1.        ]]


  1% (66 of 4541) |                       | Elapsed Time: 0:00:57 ETA:  1:06:17

connected mappoints  112
66 
 [[ 0.99999757 -0.00191782  0.00108285  0.00277477]
 [ 0.00191265  0.99998688  0.00475255  0.00742232]
 [-0.00109195 -0.00475047  0.99998812 -0.90573127]
 [ 0.          0.          0.          1.        ]]


  1% (67 of 4541) |                       | Elapsed Time: 0:00:58 ETA:  1:05:54

connected mappoints  120
67 
 [[  9.99983194e-01  -5.71734149e-03   9.61299194e-04   1.41651039e-02]
 [  5.71816798e-03   9.99983282e-01  -8.59230398e-04   1.55802660e-02]
 [ -9.56370610e-04   8.64712828e-04   9.99999169e-01  -9.00963643e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  1% (68 of 4541) |                       | Elapsed Time: 0:00:59 ETA:  1:05:28

connected mappoints  125
68 
 [[  9.99996740e-01  -2.50456530e-03   4.97364787e-04   2.69218909e-02]
 [  2.50568782e-03   9.99994286e-01  -2.26929606e-03   1.43398145e-02]
 [ -4.91678345e-04   2.27053490e-03   9.99997301e-01  -8.91242398e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  1% (69 of 4541) |                       | Elapsed Time: 0:01:00 ETA:  1:05:06

connected mappoints  126
69 
 [[ 0.99999692 -0.00222098  0.00110752  0.01066197]
 [ 0.00222258  0.99999648 -0.00144629  0.010323  ]
 [-0.0011043   0.00144875  0.99999834 -0.90369292]
 [ 0.          0.          0.          1.        ]]


  1% (70 of 4541) |                       | Elapsed Time: 0:01:01 ETA:  1:04:48

connected mappoints  132
70 
 [[  9.99999162e-01  -5.31821440e-04   1.18058232e-03   4.92800015e-03]
 [  5.31370091e-04   9.99999786e-01   3.82591112e-04   5.30350733e-03]
 [ -1.18078553e-03  -3.81963465e-04   9.99999230e-01  -8.68142910e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  1% (71 of 4541) |                       | Elapsed Time: 0:01:01 ETA:  1:04:26

connected mappoints  122
71 
 [[  9.99997342e-01  -1.57094615e-03   1.68743049e-03   5.39737440e-03]
 [  1.57250438e-03   9.99998338e-01  -9.22506032e-04   1.55061602e-02]
 [ -1.68597848e-03   9.25157072e-04   9.99998151e-01  -8.59373521e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  1% (72 of 4541) |                       | Elapsed Time: 0:01:02 ETA:  1:04:04

connected mappoints  122
72 
 [[  9.99998976e-01  -7.92918975e-04   1.19155792e-03   8.93511156e-03]
 [  7.93619136e-04   9.99999513e-01  -5.87243960e-04   1.21722741e-02]
 [ -1.19109170e-03   5.88189001e-04   9.99999118e-01  -8.39882213e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  1% (73 of 4541) |                       | Elapsed Time: 0:01:03 ETA:  1:03:35

connected mappoints  101
73 
 [[  9.99998126e-01  -1.62850629e-04   1.92886897e-03   1.43019585e-03]
 [  1.64368499e-04   9.99999677e-01  -7.86789819e-04   2.41420046e-02]
 [ -1.92874022e-03   7.87105390e-04   9.99997830e-01  -8.20483093e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  1% (74 of 4541) |                       | Elapsed Time: 0:01:04 ETA:  1:03:18

connected mappoints  99
74 
 [[  9.99999192e-01   1.49152094e-04   1.26274696e-03   3.75881888e-03]
 [ -1.47724977e-04   9.99999350e-01  -1.13018625e-03   1.57227160e-02]
 [ -1.26291471e-03   1.12999879e-03   9.99998564e-01  -7.92466290e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  1% (75 of 4541) |                       | Elapsed Time: 0:01:05 ETA:  1:03:21

connected mappoints  99
75 
 [[ 0.9999955  -0.00278276  0.00112501  0.01179788]
 [ 0.00278586  0.99999231 -0.00276179  0.01526309]
 [-0.00111731  0.00276492  0.99999555 -0.7830554 ]
 [ 0.          0.          0.          1.        ]]


  1% (76 of 4541) |                       | Elapsed Time: 0:01:05 ETA:  1:03:16

connected mappoints  118
76 
 [[ 0.99999817 -0.00154945  0.00112314  0.01081252]
 [ 0.00155417  0.99998993 -0.00421052  0.01759436]
 [-0.00111661  0.00421226  0.9999905  -0.79630581]
 [ 0.          0.          0.          1.        ]]


  1% (77 of 4541) |                       | Elapsed Time: 0:01:06 ETA:  1:03:12

connected mappoints  92
77 
 [[ 0.9999947  -0.00276048  0.00172412  0.01140453]
 [ 0.0027631   0.99999502 -0.00152191  0.0088848 ]
 [-0.00171991  0.00152666  0.99999736 -0.78375341]
 [ 0.          0.          0.          1.        ]]


  1% (78 of 4541) |                       | Elapsed Time: 0:01:07 ETA:  1:03:23

connected mappoints  114
78 
 [[ 0.99999388 -0.00320886  0.00139406  0.00854704]
 [ 0.00320251  0.99998461  0.00453115  0.00750717]
 [-0.00140858 -0.00452666  0.99998876 -0.75504374]
 [ 0.          0.          0.          1.        ]]


  1% (79 of 4541) |                       | Elapsed Time: 0:01:08 ETA:  1:03:20

connected mappoints  122
79 
 [[  9.99998973e-01  -9.62628363e-04   1.06222733e-03   1.25512223e-02]
 [  9.54798920e-04   9.99972556e-01   7.34683135e-03   1.21365209e-02]
 [ -1.06927045e-03  -7.34580959e-03   9.99972447e-01  -7.25069619e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  1% (80 of 4541) |                       | Elapsed Time: 0:01:09 ETA:  1:03:02

connected mappoints  129
80 
 [[  9.99998876e-01   5.05688818e-04   1.41161445e-03   3.65749771e-03]
 [ -5.14086048e-04   9.99982139e-01   5.95466002e-03   2.53204762e-03]
 [ -1.40857803e-03  -5.95537902e-03   9.99981275e-01  -7.14323328e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  1% (81 of 4541) |                       | Elapsed Time: 0:01:10 ETA:  1:02:48

connected mappoints  147
81 
 [[  9.99999742e-01  -6.69576036e-04   2.58845308e-04   1.46265307e-02]
 [  6.69235663e-04   9.99998914e-01   1.31282191e-03   1.71659348e-03]
 [ -2.59724061e-04  -1.31264834e-03   9.99999105e-01  -6.98056978e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  1% (82 of 4541) |                       | Elapsed Time: 0:01:11 ETA:  1:02:45

connected mappoints  132
82 
 [[  9.99997089e-01  -2.28896751e-03   7.62980905e-04   1.33986599e-02]
 [  2.29162526e-03   9.99991246e-01  -3.50090543e-03   1.00400185e-02]
 [ -7.54960768e-04   3.50264370e-03   9.99993581e-01  -6.74264619e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  1% (83 of 4541) |                       | Elapsed Time: 0:01:12 ETA:  1:02:37

connected mappoints  135
83 
 [[  9.99998920e-01   8.13054888e-05   1.46762859e-03   1.13902712e-03]
 [ -7.76304950e-05   9.99996862e-01  -2.50391862e-03   2.14992664e-02]
 [ -1.46782756e-03   2.50380199e-03   9.99995788e-01  -6.67949323e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  1% (84 of 4541) |                       | Elapsed Time: 0:01:13 ETA:  1:02:17

connected mappoints  114
84 
 [[  9.99995684e-01   2.84979391e-03   7.15165689e-04   9.45172961e-03]
 [ -2.85220612e-03   9.99990170e-01   3.39490096e-03   1.11853889e-02]
 [ -7.05483891e-04  -3.39692611e-03   9.99993982e-01  -6.37905733e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  1% (85 of 4541) |                       | Elapsed Time: 0:01:14 ETA:  1:02:12

connected mappoints  131
85 
 [[  9.99971388e-01   7.56072091e-03  -2.41752238e-04  -1.43748830e-03]
 [ -7.55939133e-03   9.99958448e-01   5.09492085e-03   1.00240311e-02]
 [  2.80263467e-04  -5.09294758e-03   9.99986992e-01  -6.14193057e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  1% (86 of 4541) |                       | Elapsed Time: 0:01:14 ETA:  1:02:04

connected mappoints  125
86 
 [[ 0.9999934   0.00338201 -0.00132996  0.00690478]
 [-0.00337562  0.99998289  0.00477667  0.00374024]
 [ 0.00134609 -0.00477215  0.99998771 -0.59987554]
 [ 0.          0.          0.          1.        ]]


  1% (87 of 4541) |                       | Elapsed Time: 0:01:15 ETA:  1:01:54

connected mappoints  136
87 
 [[  9.99992579e-01  -2.53338151e-03  -2.90229109e-03   1.94025479e-03]
 [  2.53462708e-03   9.99996697e-01   4.25567897e-04   2.37285238e-03]
 [  2.90120338e-03  -4.32920965e-04   9.99995698e-01  -5.93880696e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  1% (88 of 4541) |                       | Elapsed Time: 0:01:16 ETA:  1:01:39

connected mappoints  131
88 
 [[ 0.99997981 -0.00492303 -0.0040185   0.01043985]
 [ 0.00491079  0.99998329 -0.00304997  0.01074587]
 [ 0.00403345  0.00303017  0.99998727 -0.58423109]
 [ 0.          0.          0.          1.        ]]


  1% (89 of 4541) |                       | Elapsed Time: 0:01:17 ETA:  1:01:23

connected mappoints  133
89 
 [[ 0.99998836  0.00165375 -0.00453336 -0.00468168]
 [-0.00166785  0.99999378 -0.00310759  0.00971906]
 [ 0.00452819  0.00311511  0.9999849  -0.56609898]
 [ 0.          0.          0.          1.        ]]


  1% (90 of 4541) |                       | Elapsed Time: 0:01:18 ETA:  1:01:17

connected mappoints  147
90 
 [[ 0.99997591  0.00408376 -0.00561224 -0.0087399 ]
 [-0.00410843  0.99998192 -0.00439046  0.0153471 ]
 [ 0.00559421  0.00441341  0.99997461 -0.55448037]
 [ 0.          0.          0.          1.        ]]


  2% (91 of 4541) |                       | Elapsed Time: 0:01:18 ETA:  1:01:12

connected mappoints  130
91 
 [[ 0.99995513  0.00492419 -0.00809257 -0.01177912]
 [-0.00495413  0.99998094 -0.00368447  0.01935812]
 [ 0.00807427  0.00372439  0.99996047 -0.53509144]
 [ 0.          0.          0.          1.        ]]


  2% (92 of 4541) |                       | Elapsed Time: 0:01:19 ETA:  1:00:54

connected mappoints  129
92 
 [[  9.99938171e-01  -3.72912933e-03  -1.04760219e-02  -7.47495376e-03]
 [  3.73303989e-03   9.99992970e-01   3.53756753e-04   5.40053365e-03]
 [  1.04746291e-02  -3.92842288e-04   9.99945062e-01  -5.20859885e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  2% (93 of 4541) |                       | Elapsed Time: 0:01:20 ETA:  1:00:44

connected mappoints  127
93 
 [[  9.99870057e-01  -5.51344779e-03  -1.51483115e-02   4.12089283e-03]
 [  5.51901728e-03   9.99984717e-01   3.25884140e-04   7.46023525e-03]
 [  1.51462832e-02  -4.09445586e-04   9.99885205e-01  -5.02950031e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  2% (94 of 4541) |                       | Elapsed Time: 0:01:20 ETA:  1:00:34

connected mappoints  132
94 
 [[  9.99827061e-01  -5.34700442e-03  -1.78117427e-02  -7.73261182e-03]
 [  5.33868783e-03   9.99985617e-01  -5.14433282e-04   1.61947510e-03]
 [  1.78142371e-02   4.19252983e-04   9.99841226e-01  -5.01486490e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  2% (95 of 4541) |                       | Elapsed Time: 0:01:21 ETA:  1:00:26

connected mappoints  117
95 
 [[  9.99755776e-01   1.47322715e-03  -2.20503447e-02  -1.42502143e-02]
 [ -1.49038396e-03   9.99998599e-01  -7.61660830e-04   4.24045738e-03]
 [  2.20491917e-02   7.94338294e-04   9.99756571e-01  -4.79107651e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  2% (96 of 4541) |                       | Elapsed Time: 0:01:22 ETA:  1:00:17

connected mappoints  102
96 
 [[ 0.99965358  0.00330061 -0.02611173 -0.02367722]
 [-0.00333207  0.99999377 -0.00116138  0.00738113]
 [ 0.02610773  0.00124798  0.99965836 -0.47436345]
 [ 0.          0.          0.          1.        ]]


  2% (97 of 4541) |                       | Elapsed Time: 0:01:22 ETA:  1:00:03

connected mappoints  92
97 
 [[  9.99569787e-01   7.68063366e-04  -2.93198183e-02  -2.58978520e-02]
 [ -8.10976597e-04   9.99998617e-01  -1.45176206e-03   4.24355378e-03]
 [  2.93186627e-02   1.47491518e-03   9.99569027e-01  -4.66526094e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  2% (98 of 4541) |                       | Elapsed Time: 0:01:23 ETA:  0:59:48

connected mappoints  85
98 
 [[  9.99447943e-01   5.20934713e-05  -3.32235971e-02  -1.88252778e-02]
 [ -1.00741174e-04   9.99998925e-01  -1.46257925e-03   5.60207895e-03]
 [  3.32234852e-02   1.46511880e-03   9.99446874e-01  -4.54336487e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  2% (99 of 4541) |                       | Elapsed Time: 0:01:23 ETA:  0:59:29

connected mappoints  94
99 
 [[  9.99289352e-01  -4.45610606e-04  -3.76907485e-02  -2.40424047e-02]
 [  4.26826144e-04   9.99999781e-01  -5.06428964e-04   2.20467667e-03]
 [  3.76909659e-02   4.89981675e-04   9.99289323e-01  -4.48918369e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  2% (100 of 4541) |                      | Elapsed Time: 0:01:24 ETA:  0:59:17

connected mappoints  85
100 
 [[  9.99108890e-01  -4.10226449e-04  -4.22049535e-02  -2.56477271e-02]
 [  4.01505124e-04   9.99999896e-01  -2.15118534e-04   2.86857735e-03]
 [  4.22050374e-02   1.97981335e-04   9.99108951e-01  -4.47983464e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  2% (101 of 4541) |                      | Elapsed Time: 0:01:25 ETA:  0:59:04

connected mappoints  86
101 
 [[  9.98946115e-01   2.37147808e-04  -4.58977543e-02  -2.96216229e-02]
 [ -2.59826758e-04   9.99999847e-01  -4.88153683e-04   8.69279511e-03]
 [  4.58976315e-02   4.99564690e-04   9.98946023e-01  -4.39623022e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  2% (102 of 4541) |                      | Elapsed Time: 0:01:25 ETA:  0:58:44

connected mappoints  102
102 
 [[ 0.99872689 -0.00237217 -0.05038818 -0.0349302 ]
 [ 0.00245499  0.99999574  0.0015818   0.00523117]
 [ 0.05038421 -0.00170349  0.99872846 -0.43830455]
 [ 0.          0.          0.          1.        ]]


  2% (103 of 4541) |                      | Elapsed Time: 0:01:26 ETA:  0:58:12

connected mappoints  99
103 
 [[  9.98451654e-01  -5.70668032e-05  -5.56263584e-02  -2.70632185e-02]
 [  1.78809225e-04   9.99997600e-01   2.18359970e-03   8.92017071e-03]
 [  5.56261002e-02  -2.19016524e-03   9.98449268e-01  -4.15470156e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  2% (104 of 4541) |                      | Elapsed Time: 0:01:26 ETA:  0:57:45

connected mappoints  81
104 
 [[  9.98233351e-01   5.50596097e-05  -5.94152659e-02  -2.92081491e-02]
 [  3.32142365e-05   9.99998897e-01   1.48472124e-03   5.13119602e-03]
 [  5.94152822e-02  -1.48407170e-03   9.98232248e-01  -4.02883023e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  2% (105 of 4541) |                      | Elapsed Time: 0:01:27 ETA:  0:57:24

connected mappoints  93
105 
 [[  9.98053966e-01  -1.93416072e-04  -6.23557854e-02  -3.19718782e-02]
 [  2.72747707e-04   9.99999164e-01   1.26373230e-03   2.18802331e-03]
 [  6.23554888e-02  -1.27828044e-03   9.98053184e-01  -4.00932044e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  2% (106 of 4541) |                      | Elapsed Time: 0:01:28 ETA:  0:57:20

connected mappoints  99
106 
 [[  9.97874870e-01   3.50389216e-04  -6.51584352e-02  -3.30281623e-02]
 [ -2.68610173e-04   9.99999165e-01   1.26383608e-03   3.00046836e-03]
 [  6.51588237e-02  -1.24364805e-03   9.97874131e-01  -3.94269749e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  2% (107 of 4541) |                      | Elapsed Time: 0:01:29 ETA:  0:57:24

connected mappoints  92
107 
 [[ 0.9978175   0.00267328 -0.06597799 -0.04495167]
 [-0.002564    0.9999952   0.00174087  0.00210563]
 [ 0.06598232 -0.00156791  0.99781956 -0.39224816]
 [ 0.          0.          0.          1.        ]]


  2% (108 of 4541) |                      | Elapsed Time: 0:01:30 ETA:  0:57:44

connected mappoints  84
108 
 [[  9.97806600e-01   4.20987221e-04  -6.61952531e-02  -3.47944304e-02]
 [ -3.62380174e-04   9.99999532e-01   8.97370915e-04   1.45086860e-03]
 [  6.61955999e-02  -8.71414775e-04   9.97806285e-01  -3.84905240e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  2% (109 of 4541) |                      | Elapsed Time: 0:01:31 ETA:  0:58:15

connected mappoints  85
109 
 [[  9.97916963e-01   6.94833806e-05  -6.45114778e-02  -4.48295500e-02]
 [ -1.90121697e-04   9.99998245e-01  -1.86389169e-03   1.23178027e-02]
 [  6.45112350e-02   1.87227417e-03   9.97915224e-01  -3.77940988e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  2% (110 of 4541) |                      | Elapsed Time: 0:01:32 ETA:  0:59:01

connected mappoints  131
110 
 [[ 0.997906   -0.00210588 -0.06464658 -0.03539982]
 [ 0.00194198  0.99999474 -0.00259806  0.00416948]
 [ 0.06465172  0.00246707  0.99790484 -0.38025269]
 [ 0.          0.          0.          1.        ]]


  2% (111 of 4541) |                      | Elapsed Time: 0:01:34 ETA:  1:00:01

connected mappoints  166
111 
 [[  9.97956117e-01   7.57179888e-04  -6.38984707e-02  -3.65383471e-02]
 [ -7.78095257e-04   9.99999652e-01  -3.02437564e-04  -1.61140612e-03]
 [  6.38982195e-02   3.51538514e-04   9.97956359e-01  -3.86650059e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  2% (112 of 4541) |                      | Elapsed Time: 0:01:35 ETA:  1:00:52

connected mappoints  140
112 
 [[  9.98079848e-01   3.05927708e-03  -6.18648295e-02  -4.39528172e-02]
 [ -2.99701826e-03   9.99994905e-01   1.09913759e-03   4.09139428e-03]
 [  6.18678768e-02  -9.11617060e-04   9.98083932e-01  -3.72712564e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  2% (113 of 4541) |                      | Elapsed Time: 0:01:36 ETA:  1:01:38

connected mappoints  135
113 
 [[  9.98231921e-01   5.38899801e-03  -5.91945149e-02  -4.59263479e-02]
 [ -5.33156382e-03   9.99985151e-01   1.12815853e-03   3.09764745e-03]
 [  5.91997155e-02  -8.10564521e-04   9.98245830e-01  -3.79136758e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  2% (114 of 4541) |                      | Elapsed Time: 0:01:38 ETA:  1:02:18

connected mappoints  137
114 
 [[ 0.99839148 -0.00175444 -0.05666891 -0.0343939 ]
 [ 0.00183578  0.99999736  0.00138341  0.00150488]
 [ 0.05666633 -0.00148522  0.99839207 -0.37689504]
 [ 0.          0.          0.          1.        ]]


  2% (115 of 4541) |                      | Elapsed Time: 0:01:39 ETA:  1:02:57

connected mappoints  139
115 
 [[  9.98453995e-01  -3.92926742e-03  -5.54452905e-02  -2.57625383e-02]
 [  3.98452045e-03   9.99991669e-01   8.86020733e-04   3.80978435e-03]
 [  5.54413472e-02  -1.10557383e-03   9.98461334e-01  -3.78160534e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  2% (116 of 4541) |                      | Elapsed Time: 0:01:40 ETA:  1:03:32

connected mappoints  116
116 
 [[  9.98656761e-01  -9.41077416e-04  -5.18052917e-02  -2.64430010e-02]
 [  1.04937073e-03   9.99997321e-01   2.06323090e-03   2.07511418e-03]
 [  5.18032113e-02  -2.11482245e-03   9.98655073e-01  -3.74455026e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  2% (117 of 4541) |                      | Elapsed Time: 0:01:41 ETA:  1:04:15

connected mappoints  125
117 
 [[  9.98876569e-01   6.49445122e-03  -4.69406302e-02  -4.10260800e-02]
 [ -6.48570015e-03   9.99978910e-01   3.38733117e-04   5.90607440e-03]
 [  4.69418401e-02  -3.39097214e-05   9.98897624e-01  -3.71956852e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  2% (118 of 4541) |                      | Elapsed Time: 0:01:43 ETA:  1:05:00

connected mappoints  108
118 
 [[  9.98969368e-01   1.47722258e-02  -4.29183400e-02  -5.03352845e-02]
 [ -1.48592609e-02   9.99888134e-01  -1.70960080e-03   2.12538825e-04]
 [  4.28882843e-02   2.34557364e-03   9.99077121e-01  -3.74886484e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  2% (119 of 4541) |                      | Elapsed Time: 0:01:44 ETA:  1:05:38

connected mappoints  97
119 
 [[ 0.99928249  0.00400445 -0.03766266 -0.03800515]
 [-0.00409764  0.99998873 -0.0023976   0.00307363]
 [ 0.03765263  0.0025502   0.99928763 -0.36112528]
 [ 0.          0.          0.          1.        ]]


  2% (120 of 4541) |                      | Elapsed Time: 0:01:45 ETA:  1:06:00

connected mappoints  70
120 
 [[ 0.99939233 -0.00203633 -0.03479691 -0.0211364 ]
 [ 0.00216519  0.99999094  0.00366599  0.00129551]
 [ 0.03478913 -0.0037391   0.99938768 -0.37310591]
 [ 0.          0.          0.          1.        ]]


  2% (121 of 4541) |                      | Elapsed Time: 0:01:46 ETA:  1:06:24

connected mappoints  73
121 
 [[ 0.99940088 -0.01140323 -0.03267779  0.00261161]
 [ 0.01155997  0.99992255  0.00461167  0.00949251]
 [ 0.03262267 -0.00498666  0.9994553  -0.38449397]
 [ 0.          0.          0.          1.        ]]


  2% (122 of 4541) |                      | Elapsed Time: 0:01:47 ETA:  1:06:40

connected mappoints  64
122 
 [[ 0.9995605  -0.00494392 -0.02922968 -0.00849522]
 [ 0.00503471  0.99998273  0.0030333   0.00979981]
 [ 0.02921417 -0.00317913  0.99956812 -0.3879093 ]
 [ 0.          0.          0.          1.        ]]


  2% (123 of 4541) |                      | Elapsed Time: 0:01:48 ETA:  1:07:06

connected mappoints  91
123 
 [[  9.99655125e-01   7.38755451e-04  -2.62504516e-02  -2.06012358e-02]
 [ -6.84685040e-04   9.99997626e-01   2.06871826e-03   7.56495182e-03]
 [  2.62519176e-02  -2.05003151e-03   9.99653257e-01  -4.03779395e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  2% (124 of 4541) |                      | Elapsed Time: 0:01:50 ETA:  1:07:46

connected mappoints  103
124 
 [[ 0.99968605  0.00782573 -0.02380251 -0.0241468 ]
 [-0.00775173  0.99996484  0.00319962 -0.00205388]
 [ 0.02382671 -0.0030141   0.99971156 -0.41690785]
 [ 0.          0.          0.          1.        ]]


  2% (125 of 4541) |                      | Elapsed Time: 0:01:51 ETA:  1:08:18

connected mappoints  76
125 
 [[ 0.99977106  0.00602806 -0.02053039 -0.03030763]
 [-0.00587618  0.99995498  0.00745024 -0.00668706]
 [ 0.02057437 -0.00732789  0.99976147 -0.4062205 ]
 [ 0.          0.          0.          1.        ]]


  2% (126 of 4541) |                      | Elapsed Time: 0:01:52 ETA:  1:08:50

connected mappoints  78
126 
 [[ 0.99979838  0.00135222 -0.02003403 -0.00965393]
 [-0.00121398  0.99997539  0.00691042 -0.01209085]
 [ 0.02004288 -0.0068847   0.99977542 -0.43799225]
 [ 0.          0.          0.          1.        ]]


  2% (127 of 4541) |                      | Elapsed Time: 0:01:53 ETA:  1:09:14

connected mappoints  74
127 
 [[  9.99837851e-01  -3.39106549e-03  -1.76853923e-02  -4.01559184e-03]
 [  3.39400290e-03   9.99994231e-01   1.36080583e-04   3.74446088e-03]
 [  1.76848288e-02  -1.96082791e-04   9.99843592e-01  -4.46178034e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  2% (128 of 4541) |                      | Elapsed Time: 0:01:55 ETA:  1:09:57

connected mappoints  86
128 
 [[ 0.99987198 -0.00218993 -0.01585002  0.00282204]
 [ 0.00211657  0.99998698 -0.00464341  0.00532281]
 [ 0.01585998  0.00460927  0.9998636  -0.45495439]
 [ 0.          0.          0.          1.        ]]


  2% (129 of 4541) |                      | Elapsed Time: 0:01:56 ETA:  1:11:01

connected mappoints  107
129 
 [[  9.99896020e-01   5.11662453e-04  -1.44113682e-02  -6.05365213e-03]
 [ -5.55887993e-04   9.99995149e-01  -3.06495687e-03   3.86785666e-03]
 [  1.44097300e-02   3.07264928e-03   9.99891453e-01  -4.73265509e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  2% (130 of 4541) |                      | Elapsed Time: 0:01:58 ETA:  1:12:13

connected mappoints  93
130 
 [[ 0.99990931  0.00198857 -0.01332013 -0.01040403]
 [-0.00202039  0.99999514 -0.00237533  0.00771405]
 [ 0.01331534  0.00240203  0.99990846 -0.48059583]
 [ 0.          0.          0.          1.        ]]


  2% (131 of 4541) |                      | Elapsed Time: 0:01:59 ETA:  1:13:29

connected mappoints  95
131 
 [[  9.99926922e-01   3.27176021e-03  -1.16381772e-02  -1.00297032e-02]
 [ -3.27756124e-03   9.99994514e-01  -4.79410643e-04   3.00825852e-03]
 [  1.16365448e-02   5.17520447e-04   9.99932159e-01  -5.01936032e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  2% (132 of 4541) |                      | Elapsed Time: 0:02:01 ETA:  1:15:05

connected mappoints  112
132 
 [[  9.99942699e-01   2.95217924e-03  -1.02900061e-02  -8.15431220e-03]
 [ -2.94215755e-03   9.99995183e-01   9.88926200e-04   5.57603538e-04]
 [  1.02928760e-02  -9.58594714e-04   9.99946567e-01  -5.07966448e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  2% (133 of 4541) |                      | Elapsed Time: 0:02:03 ETA:  1:16:53

connected mappoints  153
133 
 [[  9.99968132e-01   1.67762718e-03  -7.80511411e-03  -5.36412155e-03]
 [ -1.67206179e-03   9.99998343e-01   7.19515646e-04   1.93812927e-03]
 [  7.80630825e-03  -7.06442084e-04   9.99969281e-01  -5.21688975e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  2% (134 of 4541) |                      | Elapsed Time: 0:02:05 ETA:  1:18:33

connected mappoints  149
134 
 [[  9.99982482e-01   1.64370724e-03  -5.68624834e-03  -2.44554546e-04]
 [ -1.64132206e-03   9.99998563e-01   4.24105510e-04   7.98867797e-04]
 [  5.68693728e-03  -4.14765116e-04   9.99983743e-01  -5.45729938e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  2% (135 of 4541) |                      | Elapsed Time: 0:02:06 ETA:  1:20:02

connected mappoints  127
135 
 [[  9.99982898e-01   8.53384094e-04  -5.78575470e-03   1.96154290e-03]
 [ -8.46140554e-04   9.99998855e-01   1.25429335e-03   4.48212819e-04]
 [  5.78681847e-03  -1.24937634e-03   9.99982476e-01  -5.61353453e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  2% (136 of 4541) |                      | Elapsed Time: 0:02:08 ETA:  1:21:27

connected mappoints  139
136 
 [[  9.99981478e-01   4.40354095e-04  -6.07038351e-03   2.18048142e-03]
 [ -4.36555628e-04   9.99999708e-01   6.27048430e-04   1.87393900e-03]
 [  6.07065786e-03  -6.24386755e-04   9.99981378e-01  -5.75718759e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  3% (137 of 4541) |                      | Elapsed Time: 0:02:09 ETA:  1:22:34

connected mappoints  117
137 
 [[  9.99987200e-01   2.27194160e-03  -4.52083930e-03   2.51894547e-03]
 [ -2.27092925e-03   9.99997395e-01   2.29050922e-04   9.49718320e-04]
 [  4.52134791e-03  -2.18781484e-04   9.99989755e-01  -5.94211831e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  3% (138 of 4541) |                      | Elapsed Time: 0:02:10 ETA:  1:23:29

connected mappoints  114
138 
 [[  9.99994025e-01   1.89212446e-03  -2.89300861e-03  -3.21617508e-03]
 [ -1.89184423e-03   9.99998206e-01   9.95989101e-05   2.03988733e-04]
 [  2.89319187e-03  -9.41251933e-05   9.99995810e-01  -6.15529402e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  3% (139 of 4541) |                      | Elapsed Time: 0:02:12 ETA:  1:24:21

connected mappoints  126
139 
 [[  9.99994788e-01   5.38046888e-04  -3.18347737e-03   7.96321051e-03]
 [ -5.38622282e-04   9.99999839e-01  -1.79889099e-04  -4.56002661e-03]
 [  3.18338007e-03   1.81602853e-04   9.99994917e-01  -6.47422922e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  3% (140 of 4541) |                      | Elapsed Time: 0:02:13 ETA:  1:25:08

connected mappoints  137
140 
 [[  9.99995257e-01   1.51963515e-03  -2.67911013e-03   7.91350981e-03]
 [ -1.52087261e-03   9.99998738e-01  -4.59915820e-04   2.54983548e-03]
 [  2.67840784e-03   4.63988223e-04   9.99996305e-01  -6.56459291e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  3% (141 of 4541) |                      | Elapsed Time: 0:02:14 ETA:  1:25:51

connected mappoints  128
141 
 [[  9.99988828e-01   4.64687035e-03  -8.66808863e-04   4.20467114e-03]
 [ -4.64708912e-03   9.99989171e-01  -2.50547131e-04   1.91439464e-03]
 [  8.65635216e-04   2.54572470e-04   9.99999593e-01  -6.78836317e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  3% (142 of 4541) |                      | Elapsed Time: 0:02:15 ETA:  1:26:18

connected mappoints  122
142 
 [[  9.99992331e-01   3.77444374e-03   1.04515225e-03  -3.50507912e-03]
 [ -3.77430616e-03   9.99992868e-01  -1.33578432e-04  -1.10460178e-03]
 [ -1.04564898e-03   1.29632683e-04   9.99999445e-01  -6.94028115e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  3% (143 of 4541) |                      | Elapsed Time: 0:02:15 ETA:  1:26:34

connected mappoints  119
143 
 [[  9.99996191e-01   1.60209902e-03   2.24765093e-03   6.56947410e-03]
 [ -1.60229194e-03   9.99998713e-01   8.40316240e-05   5.78265078e-04]
 [ -2.24751341e-03  -8.76326969e-05   9.99997470e-01  -7.08694521e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  3% (144 of 4541) |                      | Elapsed Time: 0:02:16 ETA:  1:26:45

connected mappoints  113
144 
 [[  9.99988746e-01   2.77149491e-03   3.85047148e-03  -2.54268752e-03]
 [ -2.77096395e-03   9.99996151e-01  -1.43220788e-04   7.25726751e-03]
 [ -3.85085359e-03   1.32549659e-04   9.99992577e-01  -7.10780502e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  3% (145 of 4541) |                      | Elapsed Time: 0:02:17 ETA:  1:26:51

connected mappoints  125
145 
 [[  9.99992190e-01   3.35513710e-04   3.93785150e-03   1.22565652e-02]
 [ -3.33565816e-04   9.99999822e-01  -4.95305427e-04   4.39683959e-03]
 [ -3.93801698e-03   4.93988026e-04   9.99992124e-01  -7.14255232e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  3% (146 of 4541) |                      | Elapsed Time: 0:02:17 ETA:  1:26:56

connected mappoints  118
146 
 [[ 0.99999045  0.00172459  0.00401517  0.00129528]
 [-0.0017183   0.99999729 -0.00156953  0.00217077]
 [-0.00401787  0.00156261  0.99999071 -0.75717028]
 [ 0.          0.          0.          1.        ]]


  3% (147 of 4541) |                      | Elapsed Time: 0:02:18 ETA:  1:27:01

connected mappoints  108
147 
 [[  9.99991563e-01  -5.38922868e-04   4.07237087e-03  -4.12298746e-04]
 [  5.45232584e-04   9.99998653e-01  -1.54844475e-03   8.14961361e-03]
 [ -4.07153089e-03   1.55065208e-03   9.99990509e-01  -7.51266947e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  3% (148 of 4541) |                      | Elapsed Time: 0:02:19 ETA:  1:26:55

connected mappoints  102
148 
 [[ 0.9999876  -0.00249361  0.00431069  0.00652192]
 [ 0.00250228  0.99999486 -0.00200685  0.00244251]
 [-0.00430566  0.00201761  0.9999887  -0.78816252]
 [ 0.          0.          0.          1.        ]]


  3% (149 of 4541) |                      | Elapsed Time: 0:02:19 ETA:  1:26:48

connected mappoints  83
149 
 [[  9.99990486e-01   4.76913685e-04   4.33602118e-03  -9.69080787e-04]
 [ -4.71219007e-04   9.99999025e-01  -1.31426857e-03   4.53997803e-03]
 [ -4.33664375e-03   1.31221285e-03   9.99989736e-01  -7.74754157e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  3% (150 of 4541) |                      | Elapsed Time: 0:02:20 ETA:  1:26:27

connected mappoints  83
150 
 [[ 0.99999092 -0.00151447  0.00398447  0.0156221 ]
 [ 0.00150986  0.99999819  0.00116024 -0.00533627]
 [-0.00398622 -0.00115421  0.99999139 -0.7906988 ]
 [ 0.          0.          0.          1.        ]]


  3% (151 of 4541) |                      | Elapsed Time: 0:02:20 ETA:  1:25:51

connected mappoints  81
151 
 [[ 0.99999335 -0.00144241  0.00334991  0.00470538]
 [ 0.0014291   0.99999109  0.00397261  0.00683815]
 [-0.00335561 -0.00396779  0.9999865  -0.80406789]
 [ 0.          0.          0.          1.        ]]


  3% (152 of 4541) |                      | Elapsed Time: 0:02:21 ETA:  1:25:03

connected mappoints  79
152 
 [[ 0.99998606 -0.0051019   0.00135951  0.01700445]
 [ 0.00509461  0.99997289  0.00531573 -0.00666851]
 [-0.00138659 -0.00530873  0.99998495 -0.79529498]
 [ 0.          0.          0.          1.        ]]


  3% (153 of 4541) |                      | Elapsed Time: 0:02:22 ETA:  1:24:09

connected mappoints  77
153 
 [[  9.99995401e-01  -3.03206726e-03   6.44060669e-05   2.89466511e-03]
 [  3.03187877e-03   9.99991626e-01   2.74877834e-03   1.11362004e-02]
 [ -7.27400084e-05  -2.74857043e-03   9.99996220e-01  -7.97881931e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  3% (154 of 4541) |                      | Elapsed Time: 0:02:22 ETA:  1:23:12

connected mappoints  75
154 
 [[  9.99992949e-01  -3.71552377e-03  -5.44719236e-04   3.17794985e-02]
 [  3.71519016e-03   9.99992911e-01  -6.12186911e-04   1.21043458e-02]
 [  5.46989970e-04   6.10158859e-04   9.99999664e-01  -8.27773155e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  3% (155 of 4541) |                      | Elapsed Time: 0:02:23 ETA:  1:22:11

connected mappoints  91
155 
 [[  9.99996087e-01  -2.79759031e-03   4.34820244e-06   9.40763591e-03]
 [  2.79759358e-03   9.99994398e-01  -1.83790557e-03   1.01818256e-02]
 [  7.93528722e-07   1.83791054e-03   9.99998311e-01  -8.18991158e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  3% (156 of 4541) |                      | Elapsed Time: 0:02:24 ETA:  1:21:26

connected mappoints  130
156 
 [[  9.99999921e-01  -2.10353146e-04  -3.36785931e-04   8.42962897e-03]
 [  2.09905540e-04   9.99999095e-01  -1.32853440e-03   1.89990513e-02]
 [  3.37065088e-04   1.32846360e-03   9.99999061e-01  -8.16625841e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  3% (157 of 4541) |                      | Elapsed Time: 0:02:25 ETA:  1:20:58

connected mappoints  100
157 
 [[  9.99997469e-01  -2.20430303e-03   4.49646144e-04   7.71583351e-03]
 [  2.20463196e-03   9.99997302e-01  -7.32360740e-04   8.56247857e-03]
 [ -4.48030585e-04   7.33350191e-04   9.99999631e-01  -8.13765411e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  3% (158 of 4541) |                      | Elapsed Time: 0:02:26 ETA:  1:20:31

connected mappoints  115
158 
 [[  9.99998776e-01  -1.51427773e-03   3.94493127e-04   1.85175386e-03]
 [  1.51418047e-03   9.99998823e-01   2.46736287e-04   9.87695378e-03]
 [ -3.94866290e-04  -2.46138651e-04   9.99999892e-01  -8.12514160e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  3% (159 of 4541) |                      | Elapsed Time: 0:02:27 ETA:  1:20:17

connected mappoints  148
159 
 [[  9.99995031e-01  -3.14874513e-03   1.50826790e-04   8.36488144e-03]
 [  3.14867653e-03   9.99994940e-01   4.52941385e-04   1.19513306e-02]
 [ -1.52252224e-04  -4.52464229e-04   9.99999886e-01  -8.14933849e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  3% (160 of 4541) |                      | Elapsed Time: 0:02:28 ETA:  1:20:12

connected mappoints  131
160 
 [[ 0.99999585 -0.00262573 -0.00118201  0.01159006]
 [ 0.00262724  0.99999574  0.00127252  0.00754006]
 [ 0.00117866 -0.00127562  0.99999849 -0.81950305]
 [ 0.          0.          0.          1.        ]]


  3% (161 of 4541) |                      | Elapsed Time: 0:02:30 ETA:  1:20:18

connected mappoints  145
161 
 [[  9.99989120e-01  -3.64941285e-03  -2.90549086e-03   1.34268094e-02]
 [  3.64716669e-03   9.99993046e-01  -7.77996326e-04   1.49273642e-02]
 [  2.90830988e-03   7.67391052e-04   9.99995476e-01  -8.23423469e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00]]


  3% (162 of 4541) |                      | Elapsed Time: 0:02:31 ETA:  1:20:21

connected mappoints  118
162 
 [[ 0.99999    -0.00219296 -0.00389706  0.01157845]
 [ 0.00218856  0.99999696 -0.00113316  0.00634924]
 [ 0.00389953  0.00112462  0.99999176 -0.82144543]
 [ 0.          0.          0.          1.        ]]


In [ ]:
DIMENSION = 400

poses = [kf.get_pose() for kf in seq.keyframes]

#### Evaluating trajectory plot

In [ ]:
def load(file):
    keyframeids, frameids, poses = load_keyframes(file)
    return keyframeids, frameids, poses.reshape(poses.shape[0], 4, 4)

sequence = 0
folder = '/home/jeroen/notebooks/urb/resultspose'
keyframeids, frameids, poses_test = load(folder + '/keyframes_%02d_all_17_1.6_1.6.npy'%(sequence))

#### Update found observations with classifications

In [ ]:
points_per_frame = []
for kf in seq.keyframes:
    kf.update_observations_per_classification()
    kf.filter_no_classification()
    obs = kf.get_observations()
    
    points_per_frame.append(obs)

#### Visualize observations with classifications per frame

In [ ]:
show(draw_observations_classification(points_per_frame[4]))

## Semantic Map

#### Plot trajectory with observations

In [ ]:
flatten = lambda l: [item for sublist in l for item in sublist]
trajectory = Trajectory(poses, (400, 400), points_per_frame)
trajectory2 = Trajectory(poses_test, (400, 400), [])

show2(trajectory.draw(True, False), trajectory2.draw(True, False))
#trajectory._poses[:10]

#### Analyze found classes for setting up a legend

In [ ]:
classifications = []
for obs in flatten(points_per_frame):
    if obs.classification not in classifications:
        classifications.append(obs.classification)

classifications

#### Nearest Neighbors

In [ ]:
from sklearn.neighbors import NearestNeighbors

coords = [p.get_world_coords() for p in flatten(points_per_frame)]
nbrs = NearestNeighbors(n_neighbors=2, algorithm='ball_tree').fit(coords)
distances, indices = nbrs.kneighbors(coords)

"indices"